File: packages/think-data-native-backend/ios/Plugin/AnalysisPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(AnalysisPlugin)
public class AnalysisPlugin: CAPPlugin {
    private let implementation = Analysis()
    
    private let ARGUMENT_TYPE_ERROR = "type error"
    @objc func initSdk(_ call: CAPPluginCall){
        let appId = call.getString("appId")
        let serverUrl = call.getString("serverUrl")
        
        guard let appId = appId else {
            call.reject(ARGUMENT_TYPE_ERROR)
            return
        }
        guard let serverUrl = serverUrl else {
            call.reject(ARGUMENT_TYPE_ERROR)
            return
        }
        
        do{
            try implementation.initSdk(appId, serverUrl)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
       
    }
    @objc func getDistinctId(_ call: CAPPluginCall){
            var result = [String:Any]()
            let id = implementation.getDistinctId()
            result["result"] = id
            call.resolve(result)
    }
    @objc func setDistinctId(_ call: CAPPluginCall){
        let id = call.getString("id")
        guard let id = id else {
            call.reject(ARGUMENT_TYPE_ERROR)
            return
        }
        do{
            try implementation.setDistinctId(id)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func signIn(_ call: CAPPluginCall){
        let id = call.getString("id")
        guard let id = id else {
            call.reject(ARGUMENT_TYPE_ERROR)
            return
        }
        do{
            try implementation.signIn(id)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func signOut(_ call: CAPPluginCall){
            implementation.signOut()
            call.resolve()
    }
    @objc func track(_ call: CAPPluginCall){
        print("aabb helloworld")
        let eventName = call.getString("eventName")
        guard let eventName = eventName else {
            call.reject(ARGUMENT_TYPE_ERROR)
            return
        }
        do{
            let m = call.options.filter {  vk in
                vk.key.hashValue != "eventName".hashValue
            }
            try implementation.track(eventName, m)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func setPublicProperties(_ call: CAPPluginCall){
        do{
            try implementation.setPublicProperties(call.options!)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func unsetPublicProperties(_ call: CAPPluginCall){
        do{
            if let key = call.getString("key") {
                try implementation.unsetPublicProperties(key)
                call.resolve()
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func clearPublicProperties(_ call: CAPPluginCall){
        do{
            try implementation.clearPublicProperties()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func getPublicProperties(_ call:CAPPluginCall){
        do{
            let vb = try implementation.getPublicProperties()
            var vmb = [String:Any]()
            vb?.forEach({ (key: AnyHashable, value: Any) in
                vmb[key as! String] = value
            })
            call.resolve(vmb)
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func eventStart(_ call:CAPPluginCall){
        let eventName = call.getString("eventName")
        do{
            guard let eventName = eventName else {
                call.reject(ARGUMENT_TYPE_ERROR)
                return
            }
            try implementation.eventStart(eventName)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func eventEnd(_ call:CAPPluginCall){
        let eventName = call.getString("eventName")
        guard let eventName = eventName else{
            call.reject(ARGUMENT_TYPE_ERROR)
            return
        }
        let properties = call.options.filter{ key in
            "eventName".hashValue != key.key.hashValue
        }
        do{
            try implementation.eventEnd(eventName, properties)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func userSet(_ call:CAPPluginCall){
        do{
            try implementation.userSet(call.options)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func userSetOnce(_ call:CAPPluginCall){
        do{
            try implementation.userSet(call.options)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func userAdd(_ call:CAPPluginCall){
        do{
            try implementation.userAdd(call.options)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func userUnSet(_ call:CAPPluginCall){
        let keys = call.getArray("keys", String.self)
        guard let keys = keys else {
            call.reject(ARGUMENT_TYPE_ERROR)
            return
        }
        do{
            try implementation.userUnSet(keys)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func userDelete(_ call:CAPPluginCall){
        do{
            try implementation.userDelete()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func userAppend(_ call:CAPPluginCall){
        do{
            var m = call.options as NSDictionary as![String:[Any]]
            try m.forEach { (key: String, value: [Any]) in
                try implementation.userAppend([key:value])
            }
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func userUniqAppend(_ call:CAPPluginCall){
        do{
            var m = call.options as NSDictionary as![String:[Any]]
            try m.forEach { (key: String, value: [Any]) in
                try implementation.userAppend([key:value])
            }
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    @objc func getDeviceId(_ call:CAPPluginCall){
        do{
            let deviceId = try implementation.getDeviceId()
            call.resolve(["result":implementation.getDeviceId()])
        } catch {
            call.reject(error.localizedDescription)
        }
    }
}

File: packages/think-data-native-backend/android/src/main/java/com/jbcz/analysis/AnalysisPlugin.kt
package com.jbcz.analysis

import android.text.TextUtils
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "Analysis")
class AnalysisPlugin : Plugin() {
    private val implementation: Analysis by lazy {
        Analysis(context)
    }

    @PluginMethod
    fun initSdk(call: PluginCall) {
        var appId = call.getString("appId", "")
        var serverUrl = call.getString("serverUrl", "")
        try {
            implementation.initSdk(appId!!, serverUrl!!)
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun getDistinctId(call: PluginCall) {
        var result = JSObject()
        try {
            result.put("result", implementation.getDistinctId())
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun setDistinctId(call: PluginCall) {
        var id = call.data.getString("id")
        if (TextUtils.isEmpty(id)) {
            call.reject("id is empty")
        } else {
            implementation.setDistinctId(id!!)
            call.resolve()
        }
    }

    @PluginMethod
    fun signIn(call: PluginCall) {
        var id = call.getString("id")
        if (TextUtils.isEmpty(id)) {
            call.reject("id is empty")
        } else {
            implementation.signIn(id!!)
            call.resolve()
        }
    }

    @PluginMethod
    fun signOut(call: PluginCall) {
        try {
            implementation.signOut()
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun track(call: PluginCall) {
        var en = call.data.getString("eventName", "")!!
        var m = call.data
        var result = jsonToMap(filterKeys = arrayOf("eventName"), jsonBody = m)
        Log.i("args", "data:${call.data},$en")
        implementation.track(eventName = en, map = result)
        call.resolve()
    }

    @PluginMethod
    fun setPublicProperties(call: PluginCall) {
        var requests = call.data
        var result = jsonToMap(jsonBody = requests)
        implementation.setPublicProperties(result)
        call.resolve()
    }
    @PluginMethod
    fun unsetPublicProperties(call:PluginCall){
        try {
            var key = call.getString("key","")
            implementation.unsetPublicProperties(key!!)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }
    @PluginMethod
    fun clearPublicProperties(call:PluginCall){
        try {
            implementation.clearPublicProperties()
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }
    @PluginMethod
    fun getPublicProperties(call:PluginCall){
        try {
            var result =JSObject()
            var map = implementation.getPublicProperties()
            map.forEach {
                result.put(it.key,it.value)
            }
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }

    }

    @PluginMethod
    fun eventStart(call: PluginCall) {
        var eventName = call.getString("eventName", "")!!
        implementation.eventStart(eventName)
        call.resolve()
    }

    @PluginMethod
    fun eventEnd(call: PluginCall) {
        var eventName = call.getString("eventName", "")!!
        var result = jsonToMap(filterKeys = arrayOf("eventName"), jsonBody = call.data)
        implementation.eventEnd(eventName, result)
        call.resolve()
    }

    @PluginMethod
    fun userSet(call: PluginCall) {
        implementation.userSet(jsonToMap(jsonBody = call.data))
        call.resolve()
    }

    @PluginMethod
    fun userSetOnce(call: PluginCall) {
        implementation.userSetOnce(jsonToMap(jsonBody = call.data))
        call.resolve()
    }

    @PluginMethod
    fun userAdd(call: PluginCall) {
        implementation.userAdd(jsonToMap(jsonBody = call.data))
        call.resolve()
    }

    @PluginMethod
    fun userUnSet(call: PluginCall) {
        try {
            var m = call.getArray("keys").toList<String>()
            implementation.userUnSet(m.toTypedArray())
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun userDelete(call: PluginCall) {
        try {
            implementation.userDelete()
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun userAppend(call: PluginCall) {
        try {
            var fmp = jsonToMap(jsonBody = call.data)
            fmp.forEach{
                implementation.userAppend(it.key,it.value)
            }
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun userUniqAppend(call: PluginCall) {
        try {
            var fmp = jsonToMap(jsonBody = call.data)
            fmp.forEach{
                implementation.userAppend(it.key,it.value)
            }
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun getDeviceId(call: PluginCall) {
        var result = JSObject()
        try {
            var id = implementation.getDeviceId()
            result.put("result", id)
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    private fun jsonToMap(filterKeys: Array<String>? = null, jsonBody: JSObject): Map<String, Any> {
        var result = HashMap<String, Any>()
        for (iter in jsonBody.keys()) {
            var isNotNone = filterKeys?.none { it == iter } ?: false
//            Not in put it or continue
            if (!isNotNone) {
                continue
            }
            result[iter] = jsonBody[iter]
        }
        return result
    }
}